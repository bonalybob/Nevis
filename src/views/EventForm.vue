<template>
  <main>
    <h1 class="mx-10 mb-1 px-1">
      <template v-if="$route.path.includes('create')">
        <back-arrow to="/events" />Create Event
      </template>
      <template v-else> <back-arrow />Update Event </template>
    </h1>
    <div v-if="$route.path.includes('create')" class="mx-12 mb-3">
      <button @click="createEvent" class="button">Create Event</button>
    </div>
    <div v-else class="mx-12 mb-3">
      <button @click="updateEvent" class="button">Update Event</button>
      <button @click="showConfirmationDialog = true" class="button">
        Delete Event
      </button>
    </div>
    <div class="my-shadow mx-12">
      <text-input v-model.trim="eventData.name" label="Name:" />
      <text-input v-model.trim="eventData.date" label="Date:" />
    </div>
    <transition name="fade">
      <confirmation-dialog
        v-if="showConfirmationDialog"
        @close="onConfirm"
        heading="Delete Event"
        message="Are You Sure You Want to Delete This Event and All Attatched Data? This Action Can't Be Recovered."
        confirm="Delete"
        cancel="Cancel"
      />
    </transition>
  </main>
</template>

<script>
import BackArrow from '@/components/BackArrow'
import TextInput from '@/components/TextInput'
import ConfirmationDialog from '@/components/ConfirmationDialog'

export default {
  components: {
    'back-arrow': BackArrow,
    'text-input': TextInput,
    'confirmation-dialog': ConfirmationDialog,
  },

  data: function() {
    return {
      showConfirmationDialog: false,
      eventData: {
        name: '',
        date: '',
        id: undefined,
      },
    }
  },

  mounted: function() {
    if (this.$database.connection === null || !this.$database.connected) {
      this.$router.push('/')
      this.$messages.addMessage('Problem Connecting To Database', 'error')
    } else if (this.$route.params && this.$route.params.id)
      this.getEventDetails()
  },

  methods: {
    getEventDetails: function() {
      return this.$database
        .query('SELECT * FROM events WHERE id=? LIMIT 1', this.$route.params.id)
        .then(result => {
          this.eventData = result[0]
        })
        .catch(() =>
          this.$messages.addMessage('Problem Fetching Event Data', 'error')
        )
    },

    createEvent: function() {
      return this.$database
        .query('INSERT INTO events SET ?', {
          name: this.eventData.name,
          date: this.eventData.date,
        })
        .then(result => this.$router.push(`/events/${result.insertId}`))
        .then(() => this.$messages.clearMessages())
        .catch(() =>
          this.$messages.addMessage('Problem Creating Event', 'error')
        )
    },

    updateEvent: function() {
      return this.$database
        .query('UPDATE events SET name=?, date=? WHERE id=?', [
          this.eventData.name,
          this.eventData.date,
          this.eventData.id,
        ])
        .then(() => this.$messages.clearMessages())
        .then(() => this.$router.push(`/events/${this.eventData.id}`))
        .catch(() =>
          this.$messages.addMessage('Problem Updating Event', 'error')
        )
    },

    deleteEvent: function() {
      return this.$database
        .query('DELETE FROM events WHERE id=?', this.eventData.id)
        .then(() => {
          this.$messages.addMessage(`Event "${this.eventData.name}" Deleted`)
          this.$router.push('/events')
        })
        .catch(() =>
          this.$messages.addMessage('Problem Deleting Event', 'error')
        )
    },

    onConfirm: function(decision) {
      this.showConfirmationDialog = false
      if (decision) this.deleteEvent()
    },
  },
}
</script>
