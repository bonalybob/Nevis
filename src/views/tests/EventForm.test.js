import { mount, shallowMount } from '@vue/test-utils'
import EventForm from '@/views/EventForm'

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - Create', () => {
  const wrapper = shallowMount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - Update', () => {
  const wrapper = shallowMount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue([{ name: 'Test', id: 12, date: '' }]) },
      $route: { params: { id: 12 }, path: '/events/edit/12' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: false, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})

test('Get Event Details - Success', async () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue([{ id: 0, name: 'Test', date: '1/2/3' }]) },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getEventDetails()
  expect(wrapper.vm.eventData).toEqual({ id: 0, name: 'Test', date: '1/2/3' })
})

test('Get Event Details - Error', async () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getEventDetails()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Fetching Event Data', 'error')
  expect(wrapper.vm.eventData.name).toBe('')
})

test('Create Event - Success', async () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue({ insertId: 12 }) },
      $route: { path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setData({ eventData: { name: 'Test', date: '' } })
  await wrapper.vm.createEvent()
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events/12')
})

test('Create Event - Error', async () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.createEvent()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Creating Event', 'error')
})

test('Update Event - Success', async () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue([{ name: 'Test', id: 12, date: '' }]) },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.updateEvent()
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events/12')
})

test('Update Event - Error', async () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.updateEvent()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Updating Event', 'error')
})

test('Delete Event - Success', async () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue([{ name: 'Test', id: 12, date: '' }]) },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.deleteEvent()
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events')
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Event "Test" Deleted')
})

test('Delete Event - Error', async () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.deleteEvent()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Deleting Event', 'error')
})

test('On Confirm', () => {
  const wrapper = mount(EventForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setMethods({ deleteEvent: jest.fn() })
  wrapper.vm.onConfirm(false)
  expect(wrapper.vm.deleteEvent).toHaveBeenCalledTimes(0)
  wrapper.vm.onConfirm(true)
  expect(wrapper.vm.deleteEvent).toHaveBeenCalledTimes(1)
})
