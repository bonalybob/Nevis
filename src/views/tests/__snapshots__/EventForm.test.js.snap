// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - Create 1`] = `
<main>
  <h1
    class="mx-10 mb-1"
  >
    <back-arrow-stub
      to=""
    />
    Update Event
    
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <button
      class="button"
    >
      Update Event
    </button>
     
    <button
      class="button"
    >
      Delete Event
    </button>
  </div>
   
  <div
    class="shadow mx-12"
  >
    <text-input-stub
      label="Name:"
      type=""
      value=""
    />
     
    <text-input-stub
      label="Date:"
      type=""
      value=""
    />
  </div>
   
  <!---->
</main>
`;

exports[`Renders Correctly - Update 1`] = `
<main>
  <h1
    class="mx-10 mb-1"
  >
    <back-arrow-stub
      to=""
    />
    Update Event
    
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <button
      class="button"
    >
      Update Event
    </button>
     
    <button
      class="button"
    >
      Delete Event
    </button>
  </div>
   
  <div
    class="shadow mx-12"
  >
    <text-input-stub
      label="Name:"
      type=""
      value=""
    />
     
    <text-input-stub
      label="Date:"
      type=""
      value=""
    />
  </div>
   
  <!---->
</main>
`;
